packages/cli/src/__tests__/lib/commands/history-util.test.ts

```diff
@@ -201,4 +211,147 @@ describe('devices-util', () => {
+		const client = { history } as SmartThingsClient
+
+		const hasNextMock = jest.fn().mockReturnValue(false)
+		const nextMock = jest.fn().mockImplementation()
+
+		const params = { locationId: 'location-id', deviceId: 'device-id' }
+		const items: DeviceActivity[] = []
+		for (let index = 0; index < maxItemsPerRequest * maxRequestsBeforeWarning + 10; index++) {
+			items.push({ deviceId: 'history-device-id', text: `item${index}` } as DeviceActivity)
+		}
+
+		const promptMock = jest.mocked(inquirer.prompt)
+
+		// Since the core SDK mutates this object via the `.next` call, each test needs its own instance.
+		const makeHistoryResponse = (items: DeviceActivity[]): PaginatedList<DeviceActivity> => ({
+			items,
+			hasNext: hasNextMock,
+			next: nextMock,
+		} as unknown as PaginatedList<DeviceActivity>)
+
+		it('uses single request when one is enough', async () => {
+			const returnedItems = items.slice(0, 2)
+			historyDevicesMock.mockResolvedValueOnce(makeHistoryResponse(returnedItems))
+
+			expect(await getHistory(client, 20, 20, params)).toStrictEqual(returnedItems)
+
+			expect(historyDevicesMock).toHaveBeenCalledTimes(1)
+			expect(historyDevicesMock).toHaveBeenCalledWith(params)
+			expect(hasNextMock).toHaveBeenCalledTimes(1)
+			expect(nextMock).toHaveBeenCalledTimes(0)
+			expect(promptMock).toHaveBeenCalledTimes(0)
+		})
+
+		it('makes multiple calls when needed', async () => {
+			const firstItemSet = items.slice(0, 2)
+			const secondItemSet = items.slice(2, 4)
+			const allItems = items.slice(0, 4)
+			const historyResponse = makeHistoryResponse(firstItemSet)
+			historyDevicesMock.mockResolvedValueOnce(historyResponse)
+			hasNextMock.mockReturnValueOnce(true)
+			nextMock.mockImplementationOnce(async () => historyResponse.items = secondItemSet)
+
+			expect(await getHistory(client, 20, 20, params)).toStrictEqual(allItems)
+
+			expect(historyDevicesMock).toHaveBeenCalledTimes(1)
+			expect(historyDevicesMock).toHaveBeenCalledWith(params)
+			expect(hasNextMock).toHaveBeenCalledTimes(2)
+			expect(nextMock).toHaveBeenCalledTimes(1)
+			expect(promptMock).toHaveBeenCalledTimes(0)
+		})
+
+		it('abandons large query when requested to do so', async () => {
+			promptMock.mockResolvedValueOnce({ answer: 'cancel' })
+
+			await expect(getHistory(client, maxItemsPerRequest * maxRequestsBeforeWarning + 1, maxItemsPerRequest,
+				params)).rejects.toThrow()
+
+			expect(promptMock).toHaveBeenCalledTimes(1)
+			expect(historyDevicesMock).toHaveBeenCalledTimes(0)
+			expect(hasNextMock).toHaveBeenCalledTimes(0)
+			expect(nextMock).toHaveBeenCalledTimes(0)
+		})
+
+		it('limits large query when requested to do so', async () => {
+			promptMock.mockResolvedValueOnce({ answer: 'reduce' })
+			const firstItemSet = items.slice(0, maxItemsPerRequest)
+			const historyResponse = makeHistoryResponse(firstItemSet)
+			historyDevicesMock.mockResolvedValueOnce(historyResponse)
+			hasNextMock.mockReturnValue(true)
+			let nextStart = 300
```

Comment:
Should this be `let nextStart = maxItemsPerRequest`? The same thing comes up in the next two tests too.

Reply:
Same answer as above.